haskell-best-practices.md

```diff
@@ -396,6 +396,37 @@ f (ThingA a) = a == 0
 f (ThingB b) = if b then 1 else 2
 ```
 
+## Avoid `OverloadedLists` and `MonoTraversable`
+
+`OverloadedLists` work from a typelcass `IsList`:
+
+```
+class IsList l where
+type Item l
+fromList  :: [Item l] -> l
+toList    :: l -> [Item l]
+```
+  
+The `mono-traversable` package defines a series typeclasses:
+
+```
+type family Element mono
+
+class MonoFoldable mono where
+ofoldMap :: Monoid m => (Element mono -> m) -> mono -> m 
+...
+```
+  
+You can see that both of these typeclasses utilize type families. `IsList` has an associated type `Item` and `MonoFoldable` uses an open type family `Element`. They are nearly identical in their utilization of these techniques. I'll focus mainly on `mono-traversable`, but arguments against it are very transferable to `OverloadedLists`.
+
+The problem is that type families do not have great type inference. The problem is compounded by the fact that `mono-traversable` supplants a large part of the prelude with functions that have bad type inference. We end up getting tons of cryptic type errors in places where we'd expect type inference to take hold. That leads to confusion and littering type annotations around the codebase. So we don't like them.
```

Comment:
> So we don't like them.

I don't know if this is necessary, especially since it's not obvious if we're talking about `mono-traversable` or type families in general :smile: